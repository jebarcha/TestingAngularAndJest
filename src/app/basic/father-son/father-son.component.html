<h5>Son</h5>
<code>{{ client | json }}</code>

<br><br>

<button 
    *ngIf="client"
    class="btn btn-outline-primary"
    (click)="onChange(2)"
    data-test="btn-id">
    Change ID
</button>

<button 
    *ngIf="client"
    class="btn btn-outline-danger"
    (click)="onDelete()"
    data-test="btn-delete">
    Delete client
</button>
