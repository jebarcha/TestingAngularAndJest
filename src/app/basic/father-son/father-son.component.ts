import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Client } from '../interfaces';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.css']
})
export class FatherSonComponent  {

  @Input() client?: Client;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onClientUpdated = new EventEmitter<Client>();

  constructor() { }

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onChange( newId: number) {
    if (!this.client) return;

    // this.client.id = newId;  // no usar
    this.client = {
      ...this.client,
      id: newId
    }
    this.onClientUpdated.emit({...this.client});
  }
  

}
