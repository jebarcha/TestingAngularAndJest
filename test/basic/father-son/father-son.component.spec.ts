import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from '../../../src/app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ FatherSonComponent ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;

    fixture.detectChanges();
    compiled = fixture.nativeElement;
    jest.clearAllMocks();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should match the snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('should not show buttons if there is no client', () => { 
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(0);
  });

  test('should show 2 buttons if there is a client', () => { 
    component.client = { id: 1, name: 'Jose'}
    fixture.detectChanges();

    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(2);
  });

  test('should match snapshot if there is a client', () => { 
    component.client = { id: 1, name: 'Jose'}
    fixture.detectChanges();

    expect(compiled).toMatchSnapshot();
  });

  test('should emit onDeleteClient with the button of Delete client', () => { 
    component.client = { id: 1, name: 'Jose'}
    fixture.detectChanges();

    jest.spyOn(component.onDeleteClient, 'emit');

    const btnDelete = compiled.querySelector('[data-test=btn-delete]');
    //console.log(btnDelete?.textContent);
    btnDelete?.dispatchEvent(new Event('click'));

    expect(component.onDeleteClient.emit).toHaveBeenCalled();

  });

  test('should emit onClientUpdated with the button of ChangeID', () => { 
    component.client = { id: 1, name: 'Jose'}
    fixture.detectChanges();

    jest.spyOn(component.onClientUpdated, 'emit');

    const btnChangeId = compiled.querySelector('[data-test=btn-id]');
    btnChangeId?.dispatchEvent(new Event('click'));

    expect(component.onClientUpdated.emit).toHaveBeenCalled();
    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      id:2,
      name: 'Jose'
    });
  });

  test('should emit onChangeClient with ID specified if there is a client', () => { 
    jest.spyOn(component.onClientUpdated, 'emit');
    component.onChange(10);
    
    expect(component.onClientUpdated.emit).not.toHaveBeenCalled();

    component.client = { id: 1, name: 'Jose'}
    fixture.detectChanges();
    component.onChange(10);

    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      id: 10,
      name: 'Jose'
    });
    
    
  });

});
