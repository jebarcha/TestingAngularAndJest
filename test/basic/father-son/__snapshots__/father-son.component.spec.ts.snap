// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent should match snapshot if there is a client 1`] = `
<div
  id="root4"
>
  <h5>
    Son
  </h5>
  <code>
    {
  "id": 1,
  "name": "Jose"
}
  </code>
  <br />
  <br />
  <button
    class="btn btn-outline-primary"
    data-test="btn-id"
  >
     Change ID

  </button>
  
  <button
    class="btn btn-outline-danger"
    data-test="btn-delete"
  >
     Delete client

  </button>
  
</div>
`;

exports[`FatherSonComponent should match the snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Son
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;
